src/shared/clients/s3Client.ts

```diff
@@ -119,13 +119,33 @@ export interface DeleteBucketRequest {
 
 const DEFAULT_CONTENT_TYPE = 'application/octet-stream'
 
+type RegionCache = {
+    [partition: string]: {
+        [name: string]: {
+            region: string
+            dates: { [region: string]: Date | undefined }
+        }
+    }
+}
+
 export class DefaultS3Client {
     public constructor(
         private readonly partitionId: string,
         private readonly regionCode: string,
         private readonly s3Provider: (regionCode: string) => Promise<S3> = createSdkClient,
         private readonly fileStreams: FileStreams = new DefaultFileStreams()
-    ) {}
+    ) {
+        DefaultS3Client.regionCache[partitionId] ??= {}
+    }
+
+    /** Cache bucket region based off partition */
+    private static regionCache: RegionCache = {}
```

Comment:
Loading S3 buckets was _painfully_ slow with 80+ buckets. It'd take 20-30 seconds to refresh because it would check the region for every bucket every single time. Not good UX. Over-aggressive caching isn't good UX either so we need to base it off partitions + the bucket create time.

Reply:
I wrote the code to handle this exact scenario (and tested it). It will evict the cache and load the correct region since the `createDate` has changed. I'll see if I can make a unit test as well (I think I wrote one but cannot remember).

Reply:
We should probably do this in all our toolkits